File: check_buttons.py
#!/usr/bin/env python3
"""
Hardware Button Verification Script for Raspberry Pi 3
-------------------------------------------------------
Checks the 2 momentary push switches wired to GPIO 23 (SW1) and GPIO 24 (SW2)
with pull-up resistors (Active-LOW logic).

Usage:
  On Raspberry Pi:
    python3 check_buttons.py

  Local simulation test (macOS / Linux / Windows):
    python3 check_buttons.py --sim
"""

import sys
import time
import argparse
from datetime import datetime

# ANSI Color formatting
GREEN = "\033[92m"
YELLOW = "\033[93m"
CYAN = "\033[96m"
MAGENTA = "\033[95m"
RED = "\033[91m"
BOLD = "\033[1m"
DIM = "\033[2m"
RESET = "\033[0m"

PIN_SW1 = 23  # GPIO 23 (Pin 16)
PIN_SW2 = 24  # GPIO 24 (Pin 18)


class ButtonChecker:
    def __init__(self, mode="auto"):
        self.mode = mode
        self.sw1_presses = 0
        self.sw2_presses = 0
        self.sw1_press_time = None
        self.sw2_press_time = None
        self.start_time = time.time()

    def get_timestamp(self):
        return datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]

    def on_press(self, sw_id, pin):
        now = time.time()
        ts = self.get_timestamp()
        if sw_id == 1:
            self.sw1_presses += 1
            self.sw1_press_time = now
            name = f"{MAGENTA}{BOLD}SW1 (GPIO {pin}){RESET}"
        else:
            self.sw2_presses += 1
            self.sw2_press_time = now
            name = f"{CYAN}{BOLD}SW2 (GPIO {pin}){RESET}"

        count = self.sw1_presses if sw_id == 1 else self.sw2_presses
        print(f"{DIM}[{ts}]{RESET} {name} {GREEN}{BOLD}▼ PRESSED {RESET} (Total: {count})")

    def on_release(self, sw_id, pin):
        now = time.time()
        ts = self.get_timestamp()
        if sw_id == 1:
            start = self.sw1_press_time
            name = f"{MAGENTA}{BOLD}SW1 (GPIO {pin}){RESET}"
        else:
            start = self.sw2_press_time
            name = f"{CYAN}{BOLD}SW2 (GPIO {pin}){RESET}"

        duration_str = ""
        if start:
            duration_ms = int((now - start) * 1000)
            duration_str = f" {CYAN}(held for {duration_ms} ms){RESET}"

        print(f"{DIM}[{ts}]{RESET} {name} {YELLOW}{BOLD}▲ RELEASED{RESET}{duration_str}")

    def run_gpiozero(self):
        """Hardware mode using modern gpiozero library."""
        from gpiozero import Button

        print(f"{CYAN}Initializing GPIO with gpiozero (pull_up=True, bounce_time=0.02s)...{RESET}")
        btn1 = Button(PIN_SW1, pull_up=True, bounce_time=0.02)
        btn2 = Button(PIN_SW2, pull_up=True, bounce_time=0.02)

        btn1.when_pressed = lambda: self.on_press(1, PIN_SW1)
        btn1.when_released = lambda: self.on_release(1, PIN_SW1)

        btn2.when_pressed = lambda: self.on_press(2, PIN_SW2)
        btn2.when_released = lambda: self.on_release(2, PIN_SW2)

        print(f"{GREEN}✓ Ready! Press the physical buttons on the breadboard.{RESET}\n")
        while True:
            time.sleep(1)

    def run_rpi_gpio(self):
        """Hardware mode fallback using RPi.GPIO library."""
        import RPi.GPIO as GPIO

        print(f"{CYAN}Initializing GPIO with RPi.GPIO (PUD_UP, bouncetime=20ms)...{RESET}")
        GPIO.setmode(GPIO.BCM)
        GPIO.setup(PIN_SW1, GPIO.IN, pull_up_down=GPIO.PUD_UP)
        GPIO.setup(PIN_SW2, GPIO.IN, pull_up_down=GPIO.PUD_UP)

        def cb_sw1(channel):
            if GPIO.input(channel) == GPIO.LOW:
                self.on_press(1, PIN_SW1)
            else:
                self.on_release(1, PIN_SW1)

        def cb_sw2(channel):
            if GPIO.input(channel) == GPIO.LOW:
                self.on_press(2, PIN_SW2)
            else:
                self.on_release(2, PIN_SW2)

        GPIO.add_event_detect(PIN_SW1, GPIO.BOTH, callback=cb_sw1, bouncetime=20)
        GPIO.add_event_detect(PIN_SW2, GPIO.BOTH, callback=cb_sw2, bouncetime=20)

        print(f"{GREEN}✓ Ready! Press the physical buttons on the breadboard.{RESET}\n")
        try:
            while True:
                time.sleep(1)
        finally:
            GPIO.cleanup()

    def run_simulated(self):
        """Local simulated mode when running on macOS/Windows/PC."""
        print(f"{YELLOW}[SIMULATION MODE]{RESET} Running simulated button pulses...")
        print(f"{DIM}Simulating button events every 3 seconds. Press Ctrl+C to stop.{RESET}\n")

        toggle = False
        while True:
            time.sleep(2.5)
            toggle = not toggle
            sw_id = 1 if toggle else 2
            pin = PIN_SW1 if sw_id == 1 else PIN_SW2

            self.on_press(sw_id, pin)
            time.sleep(0.3)
            self.on_release(sw_id, pin)

    def print_summary(self):
        uptime = int(time.time() - self.start_time)
        print("\n" + "=" * 55)
        print(f"{BOLD}Session Summary:{RESET}")
        print(f"  • Total Uptime : {uptime}s")
        print(f"  • SW1 Presses  : {self.sw1_presses}")
        print(f"  • SW2 Presses  : {self.sw2_presses}")
        print("=" * 55)


def main():
    parser = argparse.ArgumentParser(description="Check Raspberry Pi 3 push buttons (GPIO 23 & 24)")
    parser.add_argument("--sim", action="store_true", help="Force simulation mode (useful for testing on PC/Mac)")
    args = parser.parse_args()

    print("=" * 55)
    print(f"{BOLD}{CYAN}  Raspberry Pi 3 — Push Button Verification Utility{RESET}")
    print(f"{DIM}  SW1 -> GPIO 23 (Pin 16) | SW2 -> GPIO 24 (Pin 18) [Pull-Up]{RESET}")
    print("=" * 55)

    checker = ButtonChecker()

    try:
        if args.sim:
            checker.run_simulated()
        else:
            # Try gpiozero first, then RPi.GPIO, then fallback to sim
            try:
                checker.run_gpiozero()
            except ImportError:
                try:
                    checker.run_rpi_gpio()
                except (ImportError, RuntimeError) as e:
                    print(f"{YELLOW}GPIO libraries unavailable on this system ({e}).{RESET}")
                    checker.run_simulated()
    except KeyboardInterrupt:
        print(f"\n{YELLOW}Stopping button checker...{RESET}")
    finally:
        checker.print_summary()


if __name__ == "__main__":
    main()

File: src/bin/check_buttons.rs
//! Hardware Button Verification Utility for Raspberry Pi 3 (Rust port of `check_buttons.py`)
//! -------------------------------------------------------------------------------------
//! Checks the 2 momentary push switches wired to GPIO 23 (SW1) and GPIO 24 (SW2),
//! each with a 10 kΩ pull-up to 3.3 V and a 100 nF debounce cap (Active-LOW logic).
//!
//! Raspberry Pi / Linux only — there is no simulation mode. On any other platform,
//! or when the GPIO lines cannot be claimed, it exits with an explanatory message.
//!
//! Usage (on the Raspberry Pi):
//!   ./check_buttons

#[cfg(not(target_os = "linux"))]
fn main() {
    eprintln!("check_buttons requires Raspberry Pi GPIO hardware and only runs on Linux.");
    eprintln!("Build and run it on the Pi instead: make deploy-rpi && ssh <pi> ./check_buttons");
    std::process::exit(1);
}

#[cfg(target_os = "linux")]
fn main() {
    pi::main();
}

#[cfg(target_os = "linux")]
mod pi {
    use std::sync::atomic::{AtomicU32, Ordering};
    use std::sync::Arc;
    use std::time::{Duration, Instant};

    use chrono::Local;
    use colored::Colorize;
    use rppal::gpio::{Event, Gpio, InputPin, Level, Trigger};

    /// GPIO 23 (physical header pin 16)
    const PIN_SW1: u8 = 23;
    /// GPIO 24 (physical header pin 18)
    const PIN_SW2: u8 = 24;
    /// Software debounce guard. The board already filters bounce with its 10 kΩ/100 nF
    /// RC network (τ ≈ 1 ms); this only rejects anything that slips through.
    const DEBOUNCE: Duration = Duration::from_millis(20);

    // -------------------------------------------------------------------------
    // Entry point
    // -------------------------------------------------------------------------

    #[tokio::main]
    pub async fn main() {
        println!("{}", "=".repeat(55));
        println!(
            "{}",
            "  Raspberry Pi 3 — Push Button Verification Utility"
                .bright_cyan()
                .bold()
        );
        println!(
            "{}",
            "  SW1 -> GPIO 23 (Pin 16) | SW2 -> GPIO 24 (Pin 18) [Pull-Up]".dimmed()
        );
        println!("{}", "=".repeat(55));

        let stats = Arc::new(Stats::default());
        let start = Instant::now();

        if let Err(e) = start_watching(stats.clone()) {
            report_gpio_error(&e);
            std::process::exit(1);
        }

        // Block until Ctrl+C, then report the session totals and leave immediately:
        // the poll loop is parked in a blocking syscall and would stall a clean shutdown.
        let _ = tokio::signal::ctrl_c().await;
        println!("\n{}", "Stopping button checker...".bright_yellow());
        stats.print_summary(start);
        std::process::exit(0);
    }

    // -------------------------------------------------------------------------
    // GPIO monitoring
    // -------------------------------------------------------------------------

    /// Claims both lines, arms them for both-edge interrupts, and starts the poll loop.
    ///
    /// Every fallible step happens here, on the calling thread, so a failure is reported
    /// instead of leaving a silent process. (`set_async_interrupt` cannot do this: it
    /// requests the line on its own thread and returns `Ok(())` before that can fail.)
    fn start_watching(stats: Arc<Stats>) -> Result<(), rppal::gpio::Error> {
        println!("{}", "Claiming GPIO lines via rppal...".bright_cyan());

        let gpio = Gpio::new()?;
        let mut pin1 = gpio.get(PIN_SW1)?.into_input_pullup();
        let mut pin2 = gpio.get(PIN_SW2)?.into_input_pullup();

        // Idle level tells you instantly whether the pull-ups are doing their job:
        // an unpressed Active-LOW button must read High.
        report_idle_level(1, &pin1);
        report_idle_level(2, &pin2);

        // No kernel debounce attribute: the board debounces in hardware, and a rejected
        // debounce attribute is one more way for the line request to fail.
        pin1.set_interrupt(Trigger::Both, None)?;
        pin2.set_interrupt(Trigger::Both, None)?;

        println!(
            "{}\n",
            "✓ Ready! Press the physical buttons on the breadboard.".bright_green()
        );

        std::thread::spawn(move || poll_loop(gpio, pin1, pin2, stats));
        Ok(())
    }

    /// Watches both switches from a single thread.
    ///
    /// `poll_interrupts` is the supported way to wait on several pins: rppal funnels all
    /// synchronous polling through one global lock and one shared epoll, so polling each
    /// pin from its own thread makes the two switches block and cannibalise each other's
    /// events. `reset: false` returns events cached during a previous wait rather than
    /// discarding them, so simultaneous presses are both delivered.
    fn poll_loop(gpio: Gpio, pin1: InputPin, pin2: InputPin, stats: Arc<Stats>) {
        let pins = [&pin1, &pin2];
        let mut sw1 = SwitchTracker::new(1, PIN_SW1);
        let mut sw2 = SwitchTracker::new(2, PIN_SW2);

        loop {
            match gpio.poll_interrupts(&pins, false, None) {
                Ok(Some((pin, event))) => {
                    let tracker = if pin.pin() == PIN_SW1 {
                        &mut sw1
                    } else {
                        &mut sw2
                    };
                    tracker.handle(event, &stats);
                }
                Ok(None) => {} // timeout; cannot happen with an indefinite wait
                Err(e) => {
                    eprintln!("{}", format!("GPIO polling stopped: {}", e).bright_red());
                    return;
                }
            }
        }
    }

    /// Per-switch edge bookkeeping: debounce guard, press/release state, hold timing.
    struct SwitchTracker {
        sw_id: u8,
        pin_number: u8,
        pressed: bool,
        press_start: Option<Instant>,
        last_edge: Option<Instant>,
    }

    impl SwitchTracker {
        fn new(sw_id: u8, pin_number: u8) -> Self {
            Self {
                sw_id,
                pin_number,
                pressed: false,
                press_start: None,
                last_edge: None,
            }
        }

        /// Active-LOW wiring: a falling edge is a press, a rising edge is a release.
        fn handle(&mut self, event: Event, stats: &Stats) {
            let now = Instant::now();

            // Reject contact chatter, and any repeat of an edge we already acted on.
            if self.last_edge.is_some_and(|t| now - t < DEBOUNCE) {
                return;
            }
            let pressed = match event.trigger {
                Trigger::FallingEdge => true,
                Trigger::RisingEdge => false,
                _ => return,
            };
            if pressed == self.pressed {
                return;
            }

            self.last_edge = Some(now);
            self.pressed = pressed;

            if pressed {
                self.press_start = Some(now);
                stats.on_press(self.sw_id, self.pin_number);
            } else {
                let held = self.press_start.take().map_or(Duration::ZERO, |t| now - t);
                stats.on_release(self.sw_id, self.pin_number, held);
            }
        }
    }

    // -------------------------------------------------------------------------
    // Diagnostics
    // -------------------------------------------------------------------------

    fn report_idle_level(sw_id: u8, pin: &InputPin) {
        let level = pin.read();
        let note = if level == Level::High {
            "idle High — pull-up OK".bright_green()
        } else {
            "idle LOW — button held down, or the pull-up to 3.3 V is missing".bright_red()
        };
        println!("  SW{} (GPIO {}): {}", sw_id, pin.pin(), note);
    }

    fn report_gpio_error(e: &rppal::gpio::Error) {
        eprintln!("{}", format!("Cannot claim Raspberry Pi GPIO: {}", e).bright_red());
        eprintln!(
            "{}",
            "  • Is pi_node (or another GPIO program) already running and holding GPIO 23/24? \
             Stop it first — a line can only be claimed by one process."
                .dimmed()
        );
        eprintln!(
            "{}",
            "  • Otherwise check that /dev/gpiochip0 is accessible (run as root, or add the \
             user to the 'gpio' group)."
                .dimmed()
        );
    }

    // -------------------------------------------------------------------------
    // Shared press counters & output formatting
    // -------------------------------------------------------------------------

    #[derive(Default)]
    struct Stats {
        sw1_presses: AtomicU32,
        sw2_presses: AtomicU32,
    }

    impl Stats {
        fn on_press(&self, sw_id: u8, pin: u8) {
            let counter = if sw_id == 1 {
                &self.sw1_presses
            } else {
                &self.sw2_presses
            };
            let count = counter.fetch_add(1, Ordering::Relaxed) + 1;
            println!(
                "{} {} {} (Total: {})",
                timestamp().dimmed(),
                switch_label(sw_id, pin),
                "▼ PRESSED ".bright_green().bold(),
                count
            );
        }

        fn on_release(&self, sw_id: u8, pin: u8, held: Duration) {
            println!(
                "{} {} {} {}",
                timestamp().dimmed(),
                switch_label(sw_id, pin),
                "▲ RELEASED".bright_yellow().bold(),
                format!("(held for {} ms)", held.as_millis()).bright_cyan()
            );
        }

        fn print_summary(&self, start: Instant) {
            println!("\n{}", "=".repeat(55));
            println!("{}", "Session Summary:".bold());
            println!("  • Total Uptime : {}s", start.elapsed().as_secs());
            println!(
                "  • SW1 Presses  : {}",
                self.sw1_presses.load(Ordering::Relaxed)
            );
            println!(
                "  • SW2 Presses  : {}",
                self.sw2_presses.load(Ordering::Relaxed)
            );
            println!("{}", "=".repeat(55));
        }
    }

    fn timestamp() -> String {
        format!("[{}]", Local::now().format("%Y-%m-%d %H:%M:%S%.3f"))
    }

    fn switch_label(sw_id: u8, pin: u8) -> String {
        let text = format!("SW{} (GPIO {})", sw_id, pin);
        if sw_id == 1 {
            text.bright_magenta().bold().to_string()
        } else {
            text.bright_cyan().bold().to_string()
        }
    }
}
